inverse/beamformer_lcmv.m

```diff
@@ -191,143 +207,189 @@
-        [dum, maxeig]=max(diag(dd));
-        eta = vv(:,maxeig);
-        lf  = lf * eta;
-        if ~isempty(subspace), lforig = lforig * eta; end
-        dipout.ori{i} = eta;
-      otherwise
-        % compute the leadfield for the optimal dipole orientation
-        % subsequently the leadfield for only that dipole orientation will be used for the final filter computation
-        % filt = pinv(lf' * invCy * lf) * lf' * invCy;
-        % [u, s, v] = svd(real(filt * Cy * ctranspose(filt)));
-        % in this step the filter computation is not necessary, use the quick way to compute the voxel level covariance (cf. van Veen 1997)
-        [u, s, v] = svd(real(pinv(lf' * invCy *lf)));
-        eta = u(:,1);
-        lf  = lf * eta;
-        if ~isempty(subspace), lforig = lforig * eta; end
-        dipout.ori{i} = eta;
+          % and compute the leadfield for that orientation
+          lf  = lf * maxpowori;
+          dipout.ori{i} = maxpowori;
+          dipout.eta{i} = eta;
+          if hassubspace, lforig = lforig * maxpowori; end
+      end
     end
-  end
   
-  if hasfilter
-    % use the provided filter
-    filt = dip.filter{i};
-  elseif strcmp(weightnorm,'nai')
-    % Van Veen's Neural Activity Index
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(noise*filt*filt');
-    filt = pinv(sqrt(noise * lf' * invCy_squared * lf)) * lf' *invCy; % based on Sekihara & Nagarajan 2008 eqn. 4.15
-  elseif strcmp(weightnorm,'unitnoisegain')
-    % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(filt*filt');
-    filt = pinv(sqrtm(lf' * invCy_squared * lf)) * lf' *invCy;     % Sekihara & Nagarajan 2008 eqn. 4.15
-  else
     % construct the spatial filter
-    filt = pinv(lf' * invCy * lf) * lf' * invCy;              % van Veen eqn. 23, use PINV/SVD to cover rank deficient leadfield
+    switch weightnorm
+      case 'nai'
+        % Van Veen's Neural Activity Index
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(noise*filt*filt'); 
+        % the scaling term in the denominator is sqrt of projected noise, as per eqn. 2.67 of Sekihara & Nagarajan 2008 (S&N)
+        if fixedori
+          filt = pinv(sqrt(noise * lf' * invC_squared * lf)) * lf' *invC; % based on S&N eqn. 4.08
+        else
+          ft_error('vector version of nai weight normalization is not implemented');
+        end
+      case 'unitnoisegain'
+        % filt*filt' = I
+        % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(filt*filt');
```

Comment:
Curious here: have you confirmed this for the vector case? We tried in Python and could not get the output to be equivalent with normalizing the unit gain beamformer as described above (`filt_UNG = filt_UG / sqrt(filt_UG @ filt_UG.T)` ), we assume that it was something about the rotation of the output that was off, but have not looked in it further as we initially further pursued the `sqrtm()` approach over Sekihara.

Reply:
I don't know, haven't tried. Yet, I think that this inspired me to use the sqrtm. Note that currently, the vector case is dealt with if ~fixedori